### PR TITLE
Face detection with dlib

### PR DESCRIPTION
**Pull request by keogh**
Post about face detection using pytho, dlib and opencv

### CODE REVIEW

File: source/_posts/face-detection-with-dlib.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  - dlib
+---
+
+<img src="/face-detection-with-dlib/cover.jpg" />
+
+Face Detection is the first step towards Face Recognition.
+
+There are a number of algorithms to detect faces: [haar cascade](https://www.quora.com/What-is-haar-cascade), [Histogram of Oriented Gradients](https://en.wikipedia.org/wiki/Histogram_of_oriented_gradients), [Neural Networks](https://en.wikipedia.org/wiki/Artificial_neural_network). There is a [Haar implementation in OpenCV](https://docs.opencv.org/3.4.2/d7/d8b/tutorial_py_face_detection.html), and dlib uses HOG and [Neural Networks](https://towardsdatascience.com/cnn-based-face-detector-from-dlib-c3696195e01c).
+
+We'll learn how to detect faces in an image in few lines of code using python, [dlib](http://dlib.net/) and [opencv](https://opencv.org/).
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `in a few`

File: source/_posts/face-detection-with-dlib.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+<!-- more -->
+
+## The Libraries
+
+[Dlib](http://dlib.net/) is a modern C++ toolkit containing machine learning algorithms and tools for creating complex software to solve real world problems. We are gonna use its [Python API](http://dlib.net/python/index.html).
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `real-world`

Comment (hyeomans) [reply source/_posts/face-detection-with-dlib.md]: `gonna` is slang, `going to`

File: source/_posts/face-detection-with-dlib.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+[OpenCV](https://opencv.org/) is a Computer Vision library also with interfaces in C++, Python and Java, and contains a lot of functions to help us to manipulate images.
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `Python, and Java,`

File: source/_posts/face-detection-with-dlib.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+
+## Installing Dlib and Opencv
+
+Installing Python and its libraries is out of scope of this post. There are some tutorials out there that would help you get python, dlib and opencv installed. Here are a couple of links to install them on mac os:
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `Installing Python and its libraries **are out of the scope** of this post. There are some tutorials out there that would help you get python, dlib, and **OpenCV** installed. Here are a couple of links to install them on mac os`

File: source/_posts/face-detection-with-dlib.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+* https://www.learnopencv.com/install-dlib-on-macos/
+* https://www.learnopencv.com/install-opencv3-on-macos/
+
+## Face Detection
+
+In dlib's API we can find the `get_frontal_face_detector` method which "returns an object detector configured to detect human faces that are looking more or less towards the camera".
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `dlib API`
`... the camera."`

File: source/_posts/face-detection-with-dlib.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+The algorithm used to detect faces is [Histograms of Gradients (HOG)](https://lear.inrialpes.fr/people/triggs/pubs/Dalal-cvpr05.pdf) plus a [LinearSVM](https://en.wikipedia.org/wiki/Support_vector_machine).
+
+As I mentioned dlib also has a Neural Network implementation, it performs better than HOG to detect non-frontal faces, unfortunately it's computationally heavy and it's out of the scope in this post, if you are interested you could check the ["Sources"](#sources) section of this post for more info.
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `computationally heavy, and`

File: source/_posts/face-detection-with-dlib.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+In broad terms the HOG algorithm will iterate on every pixel of a given image, in each pixel it checks all the pixels around it and figures out how dark the current pixel is compared with the directly pixels surrounding it. Then it draws an arrow showing in which direction is getting darker. The process is repeated on every pixel, then it's break down in say 16x16 pixels squares where it counts up how many gradients point in each major direction, then that square is replaced with the arrow direction that was the strongest. Then using a lot of faces we could train a Linear Support Vector Machine model to detect faces in an image. - *Source:* [Machine Learning is Fun! Part 4: Modern Face Recognition with Deep Learning
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `In broad terms, the`
`in each pixel, it `
`The process is repeated on every pixel; then it breaks down in say 16x16 pixels`?
`Then using a lot of faces, we could`

File: source/_posts/face-detection-with-dlib.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+](https://medium.com/@ageitgey/machine-learning-is-fun-part-4-modern-face-recognition-with-deep-learning-c3cffc121d78)
+
+<img src="/face-detection-with-dlib/hog.png" />
+
+Dlib provides a the pretrained model using HOG through the `get_frontal_face_detector` method.
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: `Dlib provides the pre-trained model`

File: source/_posts/face-detection-with-dlib.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+## The Code
+
+We need to code the following steps:
+
+1. Import libraries.
+2. Get the dlib frontal face detector.
+3. Load the image using OpenCV.
+4. Pass de loaded image to the `detector`, it returns a array of rectangles, each rectangle containing the detected face bounds.
```
Comment (hyeomans) [inline source/_posts/face-detection-with-dlib.md]: "Pass de loaded the image to the `detector`; it returns **an** array of rectangles, each rectangle containing the detected face bounds."
is it "Pass  deloaded" or "Pass the loaded image"?

**Review by hyeomans — CHANGES_REQUESTED** [review]

File: source/_posts/face-detection-with-dlib.md (hunk 4 above)
Comment (keogh) [reply source/_posts/face-detection-with-dlib.md]: Why "OpenCV" as bold?

Comment (keogh) [reply source/_posts/face-detection-with-dlib.md]: Ahh nevermind, I see it's a correction not a markdown format change request :P

**Comment by keogh** [comment]
@hyeomans changes addressed, thanks!

**Review by hyeomans — APPROVED** [review]

**Pull request merged**